modules/standard/Lists.chpl

```diff
@@ -0,0 +1,907 @@
+        return;
+
+      const lastArrayIdx = _getLastArrayIdx();
+      const lastArrayCapacity = _getArrayCapacity(lastArrayIdx);
+      const threshhold = _totalCapacity - lastArrayCapacity;
+      const nsize = _size - amount;
+
+      if nsize >= threshhold then
+        return;
+
+      ref array = _arrays[lastArrayIdx];
+      _killArray(array);
+      _totalCapacity -= lastArrayCapacity;
+      array = nil; 
+    }
+
+    //
+    // Shift elements including and after index one to the right in memory,
+    // possibly resizing. May expand memory if necessary.
+    //
+    pragma "no doc"
+    proc _expand(idx: int) {
+      _sanity(_withinBounds(idx));
+
+      _maybeAcquireMem(1);
+
+      for i in idx.._size by -1 {
+        ref src = _getRef(i);
+        ref dst = _getRef(i + 1);
+        _move(src, dst);
+      }
+    }
+
+    //
+    // Shift all elements after index one to the left in memory, possibly
+    // resizing. This assumes the element at `idx` has already been
+    // deinitialized if that is necessary. May release memory if possible.
+    //
+    pragma "no doc"
+    proc _collapse(idx: int) {
+      _sanity(_withinBounds(idx));
+
+      if idx == _size then
+        return;
+
+      for i in idx.._size {
+        ref src = _getRef(i + 1);
+        ref dst = _getRef(i);
+        _move(src, dst);
+      }
+
+      _maybeReleaseMem(1);
+    }
+
+    //
+    // Assumes that a copy of the input element has already been made at some
+    // previous boundary, IE giving a parameter the "in" intent. Whatever 
+    // copy you've made, make sure that the "no auto destroy" pragma is
+    // attached so that you avoid firing a destructor early (and in the worst
+    // case, fire it twice).
+    //
+    pragma "no doc"
+    proc _append_by_ref(ref x: eltType) {
+      _maybeAcquireMem(1);
+      ref src = x;
+      ref dst = _getRef(_size + 1);
+      _move(src, dst);
+      _size += 1;
+    }
+
+    /*
+      Add an element to the end of this list.
+
+      :arg x: An element to append.
+    */
+    proc append(x: eltType) {
+      // TODO: Use a local copy until this pragma works with formals.
+      pragma "no auto destroy"
+      var cpy = x;
+      _enter();
+      _append_by_ref(cpy);
+      _leave();
+    }
+
+    pragma "no doc"
+    inline proc _extendGeneric(collection) {
+
+      //
+      // TODO: This could avoid repeated resizes at smaller total capacities
+      // if we resized once and then performed repeated moves, rather than
+      // calling _append().
+      //
+      for item in collection {
+        pragma "no auto destroy"
+        var cpy = item;
+        _append_by_ref(cpy);
+      }
+    }
+
+    /*
+      Extend this list by appending a copy of each element contained in
+      another list.
+
+      :arg other: A list containing elements of the same type as those
+        contained in this list.
+    */
+    proc extend(other: list(?t)) where t == eltType {
```

Comment:
Why does this use a where clause? Couldn't it just be `proc extend(other: list(eltType))` ?